### PR TITLE
Resolve Log authorization detail failure message by override ToString in each Requirement

### PR DESCRIPTION
**Pull request by wu-yafeng**
Log authorization detail failure message by override ToString in each Requirement
 - Add AuthorizationFailure parameter to internal LoggingExtensions.UserAuthorizationFailed
 - Override each Requirement ToString method

Failed Authorization Policy Details in Logs #7789

### CODE REVIEW

**Comment by HaoK** [comment]
@wu-yafeng this looks like a reasonable start, do you think you could add some tests to demonstrate what the log output would look like?

**Comment by wu-yafeng** [comment]
> @wu-yafeng this looks like a reasonable start, do you think you could add some tests to demonstrate what the log output would look like?

sure

**Comment by anurse** [comment]
@HaoK 	if we can get this reviewed an in for preview8, awesome. Otherwise it will go to 3.1

File: src/Security/Authorization/Core/src/LoggingExtensions.cs [hunk 1]
```diff
@@ -16,16 +17,22 @@ static LoggingExtensions()
...
 
         public static void UserAuthorizationSucceeded(this ILogger logger)
             => _userAuthorizationSucceeded(logger, null);
 
-        public static void UserAuthorizationFailed(this ILogger logger)
-            => _userAuthorizationFailed(logger, null);
+        public static void UserAuthorizationFailed(this ILogger logger, AuthorizationFailure failure)
+        {
+            var reason = failure.FailCalled
+                ? "Explicit called fail"
```
Comment (HaoK) [inline src/Security/Authorization/Core/src/LoggingExtensions.cs]: nit: "Fail() was explicitly called."

File: src/Security/Authorization/Core/src/LoggingExtensions.cs [hunk 2]
```diff
@@ -16,16 +17,22 @@ static LoggingExtensions()
...
+                : "Following requirement not met " + Environment.NewLine + string.Join(Environment.NewLine, failure.FailedRequirements);
```
Comment (HaoK) [inline src/Security/Authorization/Core/src/LoggingExtensions.cs]: nit: "These requirements were not met:"

File: src/Security/Authorization/Core/src/DenyAnonymousAuthorizationRequirement.cs [hunk 3]
```diff
@@ -29,5 +29,10 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
             }
             return Task.CompletedTask;
         }
+
+        public override string ToString()
+        {
+            return $"{nameof(DenyAnonymousAuthorizationRequirement)}:User authenticationted";
```
Comment (HaoK) [inline src/Security/Authorization/Core/src/DenyAnonymousAuthorizationRequirement.cs]: type: "User authenticated"?

File: src/Security/Authorization/Core/src/AssertionRequirement.cs [hunk 4]
```diff
@@ -56,5 +56,10 @@ public async Task HandleAsync(AuthorizationHandlerContext context)
                 context.Succeed(this);
             }
         }
+
+        public override string ToString()
+        {
+            return $"{nameof(AssertionRequirement)}:{nameof(Handler)} return {bool.TrueString}";
```
Comment (HaoK) [inline src/Security/Authorization/Core/src/AssertionRequirement.cs]: nit: "{nameof(Handler)} assertion should evaluate to true."

File: src/Security/Authorization/Core/src/ClaimsAuthorizationRequirement.cs [hunk 5]
```diff
@@ -69,5 +70,14 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
             }
             return Task.CompletedTask;
         }
+
+        public override string ToString()
+        {
+            var value = (AllowedValues == null || !AllowedValues.Any())
+                ? string.Empty
+                : $" AND Claim.Value in ({string.Join("|", AllowedValues)})";
```
Comment (HaoK) [inline src/Security/Authorization/Core/src/ClaimsAuthorizationRequirement.cs]: nit: " and Claim.Value is one of the following values: ({string.Join("|", AllowedValues)})"

**Review by HaoK — CHANGES_REQUESTED** [review]
Looks good except for some minor text changes

**Comment by wu-yafeng** [comment]
Poor english 😢 ,I'll fix it

**Comment by anurse** [comment]
@HaoK let's try to close this out now that we're in to the free-for-all 5.0 timeframe ;)

**Comment by HaoK** [comment]
Thanks @wu-yafeng I had to clean up some minor things and rebased, and merged these changes in PR https://github.com/aspnet/AspNetCore/pull/15350

**Pull request closed without merging**